test/cpp/util/channelz_sampler.cc

```diff
@@ -0,0 +1,518 @@
+  void StoreEntityInJson(std::string& id, std::string& type,
+                         std::string& description) {
+    std::string start, finish;
+    gpr_timespec ago = gpr_time_sub(now_, gpr_time_from_seconds(FLAGS_sampling_interval_seconds, GPR_TIMESPAN));
+    std::stringstream ss;
+    time_t tmp = GprTimespec2Timet(now_);
+    ss << std::put_time(std::localtime(&tmp), "%F %T");
+    finish = ss.str();  // example: "2019-02-01 12:12:18"
+    ss.str("");
+    tmp = GprTimespec2Timet(ago);
+    ss << std::put_time(std::localtime(&tmp), "%F %T");
+    start = ss.str();
+    grpc_core::Json obj =
+        grpc_core::Json::Object{{"Task", absl::StrFormat("%s_ID%s", type, id)},
+                                {"Start", start},
+                                {"Finish", finish},
+                                {"ID", id},
+                                {"Type", type},
+                                {"Description", description}};
+    json_.mutable_array()->push_back(obj);
+  }
+
+  // Dump data in json
+  std::string DumpJson() { return json_.Dump(); }
+
+  // Check if one entity has been recorded
+  bool CheckID(int64_t id) {
+    if (id_set_.count(id) == 0) {
+      id_set_.insert(id);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // Record current time
+  void RecordNow() {
+    now_ = gpr_now(GPR_CLOCK_REALTIME);
+  }
+
+ private:
+  std::unique_ptr<grpc::channelz::v1::Channelz::Stub> channelz_stub_;
+  std::vector<grpc::channelz::v1::Channel> top_channels_;
+  std::vector<grpc::channelz::v1::Server> all_servers_;
+  std::vector<grpc::channelz::v1::Channel> all_channels_;
+  std::vector<grpc::channelz::v1::Subchannel> all_subchannels_;
+  std::vector<grpc::channelz::v1::Socket> all_sockets_;
+  std::unordered_set<int64_t> id_set_;
+  grpc_core::Json json_;
+  int64_t rpc_timeout_seconds_;
+  gpr_timespec now_;
+  time_t GprTimespec2Timet(gpr_timespec t) {
+    time_t p = t.tv_sec;
+    return p;
+  }
+};
+
+int main(int argc, char** argv) {
+  // make sure flags can be used
+  grpc::testing::InitTest(&argc, &argv, true);
+  for (int i = 0; i < FLAGS_sampling_times; ++i) {
+    ChannelzSampler channelz_sampler;
+    channelz_sampler.Setup(FLAGS_custom_credentials_type, FLAGS_server_address);
+    std::cout << "Wait for sampling interval "
+              << FLAGS_sampling_interval_seconds << "s..." << std::endl;
+    const gpr_timespec kDelay = gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC), gpr_time_from_seconds(FLAGS_sampling_interval_seconds, GPR_TIMESPAN));
+    gpr_sleep_until(kDelay);
+    std::cout << "##### " << i << "th sampling #####" << std::endl;
+    channelz_sampler.RecordNow();
+    channelz_sampler.GetServersRPC();
+    channelz_sampler.GetSocketsOfServers();
+    channelz_sampler.GetTopChannelsRPC();
+    channelz_sampler.TraverseTopChannels();
+    channelz_sampler.DumpStdout();
+    if (!FLAGS_output_json.empty()) {
+      std::ofstream output_file(FLAGS_output_json, std::ios::app);
```

Comment:
Consider the case of running the tool multiple times, using the same command, to output to the same file. If the file is opened in append mode, then I believe they we would need to manually delete the file in between runs.

Suggest instead opening this file once, at the top of main,  without `std::ios::app`, and then leaving the file handle open, concatenating to it on each loop, and closing afterwards.

Reply:
This is a good point. Already updated. Thanks!